File: MotionPlanning/src/visualize_frames/UR5.py

import math
from robot import Robot
from link import Link
from joint import Joint, Axis
from transform import Transform 
from utilities import euler_to_quaternion
from mesh import Mesh

class UR5(Robot):
    """ """
    
    def __init__(self):
        self.mesh_file_path = "/home/pj/MotionPlanning/MotionPlanning/src/mesh_files/dae"
        super().__init__()
        
    def define_frame(self, parent_frame: str, child_frame: str, x: float, y: float, z: float, roll: float, pitch: float, yaw: float, mesh_file=None):
        """ Describe me """

        t = geometry_msgs.msg.TransformStamped()
        t.header.frame_id = parent_frame
        t.header.stamp = rospy.Time.now()
        t.child_frame_id = child_frame
        t.transform.translation.x = x
        t.transform.translation.y = y
        t.transform.translation.z = z
        t.transform.rotation = euler_to_quaternion(roll, pitch, yaw)
        return t

  
    def define_base_link(self) -> Link:
        """ """
        
        mesh_file = self.mesh_file_path + "/base.dae"
        return Link("world", "base_link", 1.0, 0.0, 0, 0, 0, 0, mesh_file)


    def define_shoulder_link(self) -> Link:
        """ """
        
        mesh_file = self.mesh_file_path + "/shoulder.dae"
        return Link("base_shoulder_joint", "shoulder_link", 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, mesh_file)

    
    def define_upper_arm_link(self) -> Link: 
        """ """

        mesh_file = self.mesh_file_path + "/upperarm.dae"
        return Link("shoulder_upper_arm_joint", "upper_arm_link", 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, mesh_file)

    def define_forearm_link(self):
        """ """

        mesh_file = self.mesh_file_path + "/forearm.dae"
        return Link("upper_arm_forearm_joint", "forearm_link", 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, mesh_file)

    def define_wrist1_link(self):
        """ """

        mesh_file = self.mesh_file_path + "/wrist1.dae"
        return Link("forearm_wrist1_joint", "wrist1_link", 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, mesh_file)

    def define_wrist2_link(self):
        """ """

        mesh_file = self.mesh_file_path + "/wrist2.dae"
        return Link("wrist1_wrist2_joint", "wrist2_link", 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, mesh_file)

    def define_wrist3_link(self):
        """ """

        mesh_file = self.mesh_file_path + "/wrist3.dae"
        return Link("wrist2_wrist3_joint", "wrist3_link", 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, mesh_file)


    def define_base_shoulder_joint(self):
        """ """

        axis = Axis([0,0,1])
        return Joint("base_link", "base_shoulder_joint", 0, 0, 0.089159, 0, 0, 0, axis, -2 * math.pi, 2 * math.pi)

    def define_shoulder_upper_arm_joint(self):
        """ """

        axis = Axis([0,1,0])
        return Joint("shoulder_link", "shoulder_upper_arm_joint", 0, 0.13585, 0, 0, math.pi/2, 0, axis, -2 * math.pi, 2 * math.pi)

    def define_upper_arm_forearm_joint(self):
        """ """

        axis = Axis([0,1,0])
        return Joint("upper_arm_link", "upper_arm_forearm_joint", 0, -0.1197, 0.42500, 0, 0, 0, axis, -2 * math.pi, 2 * math.pi)

    def define_forearm_wrist1_joint(self):
        """ """

        axis = Axis([0,1,0])
        return Joint("forearm_link", "forearm_wrist1_joint", 0, 0, 0.39225, 0, math.pi/2, 0, axis, -2 * math.pi, 2 * math.pi)

    def define_wrist1_wrist2_joint(self):
        """ """

        axis = Axis([0,0,1])
        return Joint("wrist1_link", "wrist1_wrist2_joint", 0, 0.093, 0.0, 0, 0, 0, axis, -2 * math.pi, 2 * math.pi)

    def define_wrist2_wrist3_joint(self):
        """ """

        axis = Axis([0,1,0])
        return Joint("wrist2_link", "wrist2_wrist3_joint", 0, 0.0, 0.09465, 0, 0, 0, axis, -2 * math.pi, 2 * math.pi)

    
    def get_all_joints(self):
        """ """

        base_shoulder_joint = self.define_base_shoulder_joint()
        shoulder_upper_arm_joint = self.define_shoulder_upper_arm_joint()
        upper_arm_forearm_link = self.define_upper_arm_forearm_joint()
        forearm_wrist1_joint = self.define_forearm_wrist1_joint()
        wrist1_wrist2_joint = self.define_wrist1_wrist2_joint()
        wrist2_wrist3_joint = self.define_wrist2_wrist3_joint()


        return [base_shoulder_joint, shoulder_upper_arm_joint, upper_arm_forearm_link, \
             forearm_wrist1_joint, wrist1_wrist2_joint, wrist2_wrist3_joint]
        
    def get_all_links(self):
        """ """

        base_link = self.define_base_link()
        shoulder_link = self.define_shoulder_link()
        upper_arm_link = self.define_upper_arm_link()
        forearm_link = self.define_forearm_link()
        wrist1_link = self.define_wrist1_link()
        wrist2_link = self.define_wrist2_link()
        wrist3_link = self.define_wrist3_link()

        return [base_link, shoulder_link, upper_arm_link, forearm_link, wrist1_link, wrist2_link, wrist3_link]

File: src/cpp/robot/src/UR5.cpp
#include "robots/UR5.hpp"

namespace Robot {

    Robot::UR_5::UR_5() {

        this->links = define_all_links();
        this->joints = define_all_joints();

        for (int i = 0; i < joints.size(); i++) {
            ordered_symbolic_transforms.push_back(links[i].symbolic_transform());
            ordered_symbolic_transforms.push_back(joints[i].symbolic_transform()); 
            ordered_transforms.push_back(links[i].get_transform());
            ordered_transforms.push_back(joints[i].get_transform());
        }

        ordered_symbolic_transforms.push_back(links[links.size() - 1].symbolic_transform()  );
        ordered_transforms.push_back(links[links.size() - 1].get_transform());
    }

    std::vector<geometry::Transform> UR_5::get_ordered_transforms() {

        std::vector<geometry::Transform> transforms;
        for (int i = 0; i < joints.size(); i++) {
            
            transforms.push_back(*links[i].get_transform());
            transforms.push_back(*joints[i].get_transform());
        }

        transforms.push_back(*links[links.size() - 1].get_transform());
        return transforms;
    }
    
    Robot::Link UR_5::define_base_link() {

        std::string link_name = "base_link";
        std::string visual_mesh_file = MESH_FILE_PATH_DIRECTORY + "base.dae";
        std::string collision_mesh_file = MESH_FILE_PATH_DIRECTORY + "base.stl";

        geometry::Transform root_to_base = geometry::Transform(ROOT_FRAME_NAME, link_name, 1., 0., 0., 0., 0., 0.);
        geometry::VisualMesh visual_mesh = geometry::VisualMesh(root_to_base, visual_mesh_file);

        auto parser = STLParser(collision_mesh_file);
        geometry::CollisionMesh collision_mesh = geometry::CollisionMesh(parser.get_points(), parser.get_triangle_indices());

        return Robot::Link(root_to_base, visual_mesh, collision_mesh);
    }

    Robot::Link UR_5::define_shoulder_link() {

        std::string link_name = "shoulder_link";
        std::string visual_mesh_file = MESH_FILE_PATH_DIRECTORY + "shoulder.dae";
        std::string collision_mesh_file = MESH_FILE_PATH_DIRECTORY + "shoulder.stl";

        geometry::Transform transform = geometry::Transform("base_shoulder_joint", link_name, 0., 0., 0., 0., 0., 0.);
        geometry::VisualMesh visual_mesh = geometry::VisualMesh(transform, visual_mesh_file);

        auto parser = STLParser(collision_mesh_file);
        geometry::CollisionMesh collision_mesh = geometry::CollisionMesh(parser.get_points(), parser.get_triangle_indices());

        return Link(transform, visual_mesh, collision_mesh);
    }
    
    Robot::Link UR_5::define_upper_arm_link() {

        std::string link_name = "upper_arm_link";
        std::string visual_mesh_file = MESH_FILE_PATH_DIRECTORY + "upperarm.dae";
        std::string collision_mesh_file = MESH_FILE_PATH_DIRECTORY + "upperarm.stl";;

        geometry::Transform transform = geometry::Transform("shoulder_upper_arm_joint", link_name, 0., 0., 0.0, 0., 0., 0.);
        geometry::VisualMesh visual_mesh = geometry::VisualMesh(transform, visual_mesh_file);
        auto parser = STLParser(collision_mesh_file);
        geometry::CollisionMesh collision_mesh = geometry::CollisionMesh(parser.get_points(), parser.get_triangle_indices());

        return Link(transform, visual_mesh, collision_mesh);
    }

    Robot::Link UR_5::define_forearm_link() {

        std::string link_name = "forearm_link";
        std::string visual_mesh_file = MESH_FILE_PATH_DIRECTORY + "forearm.dae";
        std::string collision_mesh_file = MESH_FILE_PATH_DIRECTORY + "forearm.stl";

        geometry::Transform transform = geometry::Transform("upper_arm_forearm_joint", link_name, 0., 0., 0., 0., 0., 0.);
        geometry::VisualMesh visual_mesh = geometry::VisualMesh(transform, visual_mesh_file);
        auto parser = STLParser(collision_mesh_file);
        geometry::CollisionMesh collision_mesh = geometry::CollisionMesh(parser.get_points(), parser.get_triangle_indices());

        return Link(transform, visual_mesh, collision_mesh);
    }
    
    Robot::Link UR_5::define_wrist1_link() {

        std::string link_name = "wrist1_link";
        std::string visual_mesh_file = MESH_FILE_PATH_DIRECTORY + "wrist1.dae";
        std::string collision_mesh_file = MESH_FILE_PATH_DIRECTORY + "wrist1.stl";

        geometry::Transform transform = geometry::Transform("forearm_wrist1_joint", link_name, 0., 0., 0., 0., 0., 0.);
        geometry::VisualMesh visual_mesh = geometry::VisualMesh(transform, visual_mesh_file);
        auto parser = STLParser(collision_mesh_file);
        geometry::CollisionMesh collision_mesh = geometry::CollisionMesh(parser.get_points(), parser.get_triangle_indices());
        
        return Link(transform, visual_mesh, collision_mesh);
    }

    Robot::Link UR_5::define_wrist2_link() {

        std::string link_name = "wrist2_link";
        std::string visual_mesh_file = MESH_FILE_PATH_DIRECTORY + "wrist2.dae";
        std::string collision_mesh_file = MESH_FILE_PATH_DIRECTORY + "wrist2.stl";

        geometry::Transform transform = geometry::Transform("wrist1_wrist2_joint", link_name, 0., 0., 0., 0., 0., 0.);
        geometry::VisualMesh visual_mesh = geometry::VisualMesh(transform, visual_mesh_file);
        auto parser = STLParser(collision_mesh_file);
        geometry::CollisionMesh collision_mesh = geometry::CollisionMesh(parser.get_points(), parser.get_triangle_indices());
        
        return Link(transform, visual_mesh, collision_mesh);
    }

    Robot::Link UR_5::define_wrist3_link() {

        std::string link_name = "wrist3_link";
        std::string visual_mesh_file = MESH_FILE_PATH_DIRECTORY + "wrist3.dae";
        std::string collision_mesh_file = MESH_FILE_PATH_DIRECTORY + "wrist3.stl";;

        geometry::Transform transform = geometry::Transform("wrist2_wrist3_joint", link_name, 0., 0., 0., 0., 0., 0.);
        geometry::VisualMesh visual_mesh = geometry::VisualMesh(transform, visual_mesh_file);
        auto parser = STLParser(collision_mesh_file);
        geometry::CollisionMesh collision_mesh = geometry::CollisionMesh(parser.get_points(), parser.get_triangle_indices());

        return Link(transform, visual_mesh, collision_mesh);
    }

    Robot::Joint UR_5::define_base_shoulder_joint() {

        geometry::Transform transform = geometry::Transform("base_link", "base_shoulder_joint", 0., 0., 0.089159, 0., 0., 0.);
        geometry::Axis axis = geometry::Axis(geometry::Vector3D(0, 0, 1));
        return Joint(transform, axis);
    }

    Robot::Joint UR_5::define_shoulder_upper_arm_joint() {

        geometry::Transform transform = geometry::Transform("shoulder_link", "shoulder_upper_arm_joint", 0., 0.13585, 0., 0., M_PI/2, 0.);
        geometry::Axis axis = geometry::Axis(geometry::Vector3D(0, 1, 0));
        return Joint(transform, axis);
    }
    
    Robot::Joint UR_5::define_upper_arm_forearm_joint() {

        geometry::Transform transform = geometry::Transform("upper_arm_link", "upper_arm_forearm_joint", 0., -0.1197 + -0.001, 0.42500, 0., 0., 0.);
        geometry::Axis axis = geometry::Axis(geometry::Vector3D(0, 1, 0));
        return Joint(transform, axis);
    }
    
    Robot::Joint UR_5::define_forearm_wrist1_joint() {

        geometry::Transform transform = geometry::Transform("forearm_link", "forearm_wrist1_joint", 0., 0.001, 0.39225, 0., M_PI/2, 0.);
        geometry::Axis axis = geometry::Axis(geometry::Vector3D(0, 1, 0));
        return Joint(transform, axis);
    }
    
    Robot::Joint UR_5::define_wrist1_wrist2_joint() {

        geometry::Transform transform = geometry::Transform("wrist1_link", "wrist1_wrist2_joint", 0., 0.093, 0., 0., 0., 0.);
        geometry::Axis axis = geometry::Axis(geometry::Vector3D(0, 0, 1));
        return Joint(transform, axis);
    }
    
    Robot::Joint UR_5::define_wrist2_wrist3_joint() {

        geometry::Transform transform = geometry::Transform("wrist2_link", "wrist2_wrist3_joint", 0., 0.001, 0.09465, 0., 0., 0.);
        geometry::Axis axis = geometry::Axis(geometry::Vector3D(0, 1, 0));
        return Joint(transform, axis);
    }

    std::vector<Joint> UR_5::define_all_joints() {

        return {define_base_shoulder_joint(), define_shoulder_upper_arm_joint(), define_upper_arm_forearm_joint(), define_forearm_wrist1_joint(), define_wrist1_wrist2_joint(), define_wrist2_wrist3_joint() };
    }
    
    std::vector<Link> UR_5::define_all_links() {

        return {define_base_link(), define_shoulder_link(), define_upper_arm_link(), define_forearm_link(), define_wrist1_link(), define_wrist2_link(), define_wrist3_link()};
    }
}